--- a/drivers/net/wireless/ath/ath10k/core.c
+++ b/drivers/net/wireless/ath/ath10k/core.c
@@ -662,6 +662,42 @@ static const struct ath10k_hw_params ath
 		.delay_unmap_buffer = false,
 		.mcast_frame_registration = false,
 	},
+        {
+               .id = QCA9377_HW_1_1_DEV_VERSION,
+               .dev_id = QCA9377_1_1_DEVICE_ID,
+               .bus = ATH10K_BUS_USB,
+               .name = "qca9377 hw1.1 usb",
+               .patch_load_addr = QCA9377_HW_1_1_PATCH_LOAD_ADDR,
+               .uart_pin = 19,
+               .led_pin = 0,
+               .otp_exe_param = 0,
+               .channel_counters_freq_hz = 88000,
+               .max_probe_resp_desc_thres = 0,
+               .cal_data_len = 8124,
+               .fw = {
+                       .dir = QCA9377_HW_1_1_FW_DIR,
+                       .board_size = QCA9377_BOARD_DATA_SZ,
+                       .board_ext_size = QCA9377_BOARD_EXT_DATA_SZ,
+               },
+               .rx_desc_ops = &qca988x_rx_desc_ops,
+               .hw_ops = &qca6174_ops,
+               .hw_clk = qca6174_clk,
+               .target_cpu_freq = 176000000,
+               .decap_align_bytes = 4,
+               .n_cipher_suites = 8,
+               .num_peers = TARGET_QCA9377_HL_NUM_PEERS,
+               .ast_skid_limit = 0x10,
+               .num_wds_entries = 0x20,
+               .uart_pin_workaround = true,
+               .start_once = true,
+               .tx_chain_mask = 0x3,
+               .rx_chain_mask = 0x3,
+               .dynamic_sar_support = false,
+               .hw_restart_disconnect = false,
+               .use_fw_tx_credits = true,
+               .delay_unmap_buffer = false,
+               .mcast_frame_registration = false,
+        },
 	{
 		.id = QCA4019_HW_1_0_DEV_VERSION,
 		.dev_id = 0,
@@ -2938,6 +2974,9 @@ int ath10k_core_start(struct ath10k *ar,
 	int status;
 	u32 val;
 
+	if (ar->is_started && ar->hw_params.start_once)
+		return 0;
+
 	lockdep_assert_held(&ar->conf_mutex);
 
 	clear_bit(ATH10K_FLAG_CRASH_FLUSH, &ar->dev_flags);
@@ -3246,6 +3285,8 @@ int ath10k_core_start(struct ath10k *ar,
 	if (status)
 		goto err_hif_stop;
 
+	ar->is_started = true;
+
 	return 0;
 
 err_hif_stop:
@@ -3300,6 +3341,8 @@ void ath10k_core_stop(struct ath10k *ar)
 	ath10k_wmi_detach(ar);
 
 	ar->id.bmi_ids_valid = false;
+
+	ar->is_started = false;
 }
 EXPORT_SYMBOL(ath10k_core_stop);
 
@@ -3441,12 +3484,16 @@ static int ath10k_core_probe_fw(struct a
 		goto err_unlock;
 	}
 
-	ath10k_debug_print_boot_info(ar);
-	ath10k_core_stop(ar);
-
-	mutex_unlock(&ar->conf_mutex);
+        /* Leave target running if hw_params.start_once is set */
+        if (ar->hw_params.start_once) {
+                mutex_unlock(&ar->conf_mutex);
+        } else {
+                ath10k_debug_print_boot_info(ar);
+                ath10k_core_stop(ar);
+                mutex_unlock(&ar->conf_mutex);
+                ath10k_hif_power_down(ar);
+        }
 
-	ath10k_hif_power_down(ar);
 	return 0;
 
 err_unlock:
--- a/drivers/net/wireless/ath/ath10k/core.h
+++ b/drivers/net/wireless/ath/ath10k/core.h
@@ -1311,6 +1311,7 @@ struct ath10k {
 
 	bool coex_support;
 	int coex_gpio_pin;
+	bool is_started;
 
 	s32 tx_power_2g_limit;
 	s32 tx_power_5g_limit;
--- a/drivers/net/wireless/ath/ath10k/hw.h
+++ b/drivers/net/wireless/ath/ath10k/hw.h
@@ -26,6 +26,7 @@ enum ath10k_bus {
 #define QCA6164_2_1_DEVICE_ID   (0x0041)
 #define QCA6174_2_1_DEVICE_ID   (0x003e)
 #define QCA6174_3_2_DEVICE_ID   (0x0042)
+#define QCA9377_1_1_DEVICE_ID   (0x9378)
 #define QCA99X0_2_0_DEVICE_ID   (0x0040)
 #define QCA9888_2_0_DEVICE_ID	(0x0056)
 #define QCA9984_1_0_DEVICE_ID	(0x0046)
@@ -115,6 +116,14 @@ enum qca9377_chip_id_rev {
 #define QCA9377_HW_1_0_FW_DIR          ATH10K_FW_DIR "/QCA9377/hw1.0"
 #define QCA9377_HW_1_0_PATCH_LOAD_ADDR	0x1234
 
+/* QCA9377 1.1 definitions */
+#define QCA9377_HW_1_1_DEV_VERSION     0x05020001
+#define QCA9377_HW_1_1_FW_DIR          ATH10K_FW_DIR "/QCA9377/hw1.1"
+#define QCA9377_HW_1_1_BOARD_DATA_FILE "board.bin"
+#define QCA9377_HW_1_1_SDIO_BOARD_DATA_FILE "board-sdio.bin"
+#define QCA9377_HW_1_1_USB_BOARD_DATA_FILE "board-usb.bin"
+#define QCA9377_HW_1_1_PATCH_LOAD_ADDR 0x1234
+
 /* QCA4019 1.0 definitions */
 #define QCA4019_HW_1_0_DEV_VERSION     0x01000000
 #define QCA4019_HW_1_0_FW_DIR          ATH10K_FW_DIR "/QCA4019/hw1.0"
@@ -634,6 +643,8 @@ struct ath10k_hw_params {
 
 	/* The hardware support multicast frame registrations */
 	bool mcast_frame_registration;
+
+	bool start_once;
 };
 
 struct htt_resp;
--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -4827,8 +4827,12 @@ void ath10k_halt(struct ath10k *ar)
 	ath10k_scan_finish(ar);
 	ath10k_peer_cleanup_all(ar);
 	ath10k_stop_radar_confirmation(ar);
-	ath10k_core_stop(ar);
-	ath10k_hif_power_down(ar);
+
+        /* Leave target running if hw_params.start_once is set */
+        if (!ar->hw_params.start_once) {
+                ath10k_core_stop(ar);
+                ath10k_hif_power_down(ar);
+        }
 
 	spin_lock_bh(&ar->data_lock);
 	list_for_each_entry(arvif, &ar->arvifs, list)
--- a/drivers/net/wireless/ath/ath10k/pci.c
+++ b/drivers/net/wireless/ath/ath10k/pci.c
@@ -3867,3 +3867,5 @@ MODULE_FIRMWARE(QCA6174_HW_3_0_FW_DIR "/
 MODULE_FIRMWARE(QCA9377_HW_1_0_FW_DIR "/" ATH10K_FW_API6_FILE);
 MODULE_FIRMWARE(QCA9377_HW_1_0_FW_DIR "/" ATH10K_FW_API5_FILE);
 MODULE_FIRMWARE(QCA9377_HW_1_0_FW_DIR "/" ATH10K_BOARD_DATA_FILE);
+MODULE_FIRMWARE(QCA9377_HW_1_1_FW_DIR "/" QCA9377_HW_1_1_SDIO_BOARD_DATA_FILE);
+MODULE_FIRMWARE(QCA9377_HW_1_1_FW_DIR "/" QCA9377_HW_1_1_USB_BOARD_DATA_FILE);
--- a/drivers/net/wireless/ath/ath10k/usb.c
+++ b/drivers/net/wireless/ath/ath10k/usb.c
@@ -1107,6 +1107,7 @@ static int ath10k_usb_pm_resume(struct u
 /* table of devices that work with this driver */
 static struct usb_device_id ath10k_usb_ids[] = {
 	{USB_DEVICE(0x13b1, 0x0042)}, /* Linksys WUSB6100M */
+	{USB_DEVICE(0x0CF3, 0x9378)}, /* Qualcomm QCA9377-7 */
 	{ /* Terminating entry */ },
 };
 
